### PR TITLE
BUG URL quote branch names and filepaths in Binder URLs

### PR DESCRIPTION
**Pull request by sdhiscocks**
Fixes issue with special characters in Binder URLs for branch names and file paths. No characters that need quoting are allowed for GitHub _orgs_ and _repos_

### CODE REVIEW

**Comment by lucyleeow** [comment]
The changes look good. I think the error is related to myavi build @larsoner ?
```
Building wheels for collected packages: mayavi, Sphinx, pyface, traitsui, memory-profiler, pypandoc, apptools, configobj, envisage
  Building wheel for mayavi (setup.py): started
  Building wheel for mayavi (setup.py): finished with status 'error'
  ERROR: Command errored out with exit status 1:
   command: /usr/share/miniconda/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-goy0uugo/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-goy0uugo/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-a85htd_0
       cwd: /tmp/pip-req-build-goy0uugo/
  Complete output (33 lines):
  running bdist_wheel
  running build
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/pip-req-build-goy0uugo/setup.py", line 411, in <module>
      setup(
    File "/usr/share/miniconda/lib/python3.8/site-packages/numpy/distutils/core.py", line 169, in setup
      return old_setup(**new_attr)
    File "/usr/share/miniconda/lib/python3.8/site-packages/setuptools/__init__.py", line 153, in setup
      return distutils.core.setup(**attrs)
  ModuleNotFoundError: No module named 'vtk'
  ----------------------------------------
  ERROR: Failed building wheel for mayavi
  Running setup.py clean for mayavi
```

**Comment by larsoner** [comment]
To fix CIs for now I would pin `vtk<=9.0.1` in some requirements. 9.0.2/9.0.3 cause problems and Mayavi needs to be updated to at least build (though it's still broken on Windows even with this PR):

https://github.com/enthought/mayavi/pull/1050

File: sphinx_gallery/gen_rst.py [hunk 1]
```diff
@@ -542,9 +542,149 @@ def _get_memory_base(gallery_conf):
     return memory_base
 
 
-def execute_code_block(compiler, block, example_globals,
-                       script_vars, gallery_conf):
-    """Execute the code block of the example file."""
+def _ast_module():
```
Comment (larsoner) [inline sphinx_gallery/gen_rst.py:545]: @lucyleeow something is wrong with the diff, this stuff shouldn't be in here

**Comment by lucyleeow** [comment]
Not sure what happened there but fixed now I think. Thanks @larsoner

**Comment by lucyleeow** [comment]
Thanks @sdhiscocks !

**Pull request merged**